lib/render/FlowRenderer.cpp

```diff
@@ -322,17 +345,257 @@ FlowRenderer::_paintGL( bool fast )
+                        continue;
+                    }
+                    
+                    vec3 prep(-normalize(sv[1].p-sv[0].p) + sv[0].p);
+                    vec3 post( normalize(sv[svn-1].p-sv[svn-2].p) + sv[svn-1].p);
+                    
+                    size_t vn = vertices.size();
+                    vertices.resize(vn + svn + 2);
+                    vertices[vn] = {prep, sv[0].v};
+                    vertices[vertices.size()-1] = {post, sv[svn-1].v};
+                    
+                    memcpy(vertices.data() + vn + 1, sv.data(), sizeof(Vertex) * svn);
+                    
+                    sizes.push_back(svn+2);
+                    sv.clear();
+                } else {
+                    const flow::Particle &p = stream[i];
+                    
+                    if (_cache_isSteady) {
+                        sv.push_back({p.location, p.value});
+                    } else {
+                        if(p.time > _timestamps.at(_cache_currentTS))
+                            continue;
+                        if(p.time >= startingTime)
+                            sv.push_back({p.location, p.value});
+                    }
+                }
+            }
+            
+            _renderStatus = FlowStatus::UPTODATE;
+        }
+
+        assert(glIsVertexArray(_VAO) == GL_TRUE);
+        assert(glIsBuffer(_VBO) == GL_TRUE);
+
+        glBindVertexArray(_VAO);
+        glBindBuffer(GL_ARRAY_BUFFER, _VBO);
+        glBufferData(GL_ARRAY_BUFFER, sizeof(Vertex) * vertices.size(), vertices.data(), GL_STREAM_DRAW);
+        glBindBuffer(GL_ARRAY_BUFFER, 0);
+        
+        _streamSizes = sizes;
+    }
+    
+    
+    FlowParams::RenderType renderType = (FlowParams::RenderType)rp->GetValueLong(FlowParams::RenderTypeTag, FlowParams::RenderTypeStream);
+    FlowParams::GlpyhType glyphType = (FlowParams::GlpyhType)rp->GetValueLong(FlowParams::RenderGlyphTypeTag, FlowParams::GlpyhTypeSphere);
+    bool geom3d = rp->GetValueLong(FlowParams::RenderGeom3DTag, false);
+    bool show_dir = rp->GetValueLong(FlowParams::RenderShowStreamDirTag, false);
+    
+    float radiusBase = rp->GetValueDouble(FlowParams::RenderRadiusBaseTag, -1);
+    if (radiusBase == -1) {
+        vector<double> mind, maxd;
+        _dataMgr->GetVariableExtents(rp->GetCurrentTimestep(), rp->GetColorMapVariableName(), rp->GetRefinementLevel(), rp->GetCompressionLevel(), mind, maxd);
+        vec3 min(mind[0], mind[1], mind[2]);
+        vec3 max(maxd[0], maxd[1], maxd[2]);
+        vec3 lens = max - min;
+        float largestDim  = glm::max(lens.x, glm::max(lens.y, lens.z));
+        radiusBase = largestDim / 560.f;
+        rp->SetValueDouble(FlowParams::RenderRadiusBaseTag, "", radiusBase);
+    }
+    float radiusScalar = rp->GetValueDouble(FlowParams::RenderRadiusScalarTag, 1);
+    float radius = radiusBase * radiusScalar;
+    int   glyphStride = rp->GetValueLong(FlowParams::RenderGlyphStrideTag, 5);
+    
+show_dir_loop:
```

Comment:
As goto is considered an uncommon logic flow, can you document what it does and why it is necessary? 

For example, my question is, Line 572-575 seem to indicate that variable `show_dir` is set to false right after show_dir_loop is executed. In that case, why not put show_dir_loop in a if conditioned branch?

Reply:
It is just a different type of loop which repeats the rendering with show_dir disabled. I'm not sure what you mean by "put show_dir_loop in a if conditioned branch".

Reply:
I meant if there's a reasonable alternative to `goto`? For example a regular if branch or things like that?

Reply:
Sure. All the goto does is reruns a portion of code with a different parameter in a special case. There are many ways to do this but I felt that the goto best represented the intention.

Reply:
In that case, are you able to adopt a conditional instead of a goto? 

The main reason is that goto isn't very commonly used, and it certainly doesn't make the code easier to understand. There are very few cases goto is considered a better choice than others, but I'm afraid this isn't one of those cases.

More comprehensive arguments against using goto is available on the internet, which I'm sure you can easily find.

Reply:
I'm with Sam on this one. The control flow is not particularly clear. Couldn't you simply put code after the label in a helper function?

Reply:
Done.